### PR TITLE
Add descheduler annotations support

### PR DESCRIPTION
**Pull request by fossedihelm**
### What this PR does
Add the descheduler annotations, allowing the compatibility with the [Descheduler](https://github.com/kubernetes-sigs/descheduler)

#### Before this PR:
There is no compatibility with the descheduler since the way KubeVirt handles eviction requests causes the descheduler to make wrong decisions and take wrong actions that could destabilize the cluster.

#### After this PR:
KubeVirt adds the proper annotations to the virt-launcher pods so that the descheduler is correctly instructed on handling the eviction response and the virt-launcher pods.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 
Design proposal:
- https://github.com/kubevirt/community/pull/258
- https://github.com/kubevirt/community/pull/295

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (https://github.com/kubevirt/community/pull/258,https://github.com/kubevirt/community/pull/295 ) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [x] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [x] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
Add descheduler compatibility
```

### CODE REVIEW

File: pkg/virt-controller/watch/vmi.go [hunk 1]
```diff
@@ -1271,7 +1272,9 @@ func (c *VMIController) sync(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod,
 	}
 
 	if !isTempPod(pod) && isPodReady(pod) {
-		newAnnotations := network.GeneratePodAnnotations(vmi.Spec.Networks, vmi.Spec.Domain.Devices.Interfaces, pod.Annotations[networkv1.NetworkStatusAnnot], c.clusterConfig.GetNetworkBindings())
+		newAnnotations := map[string]string{virtv1.DeschedulerRequestEvictOnlyAnnotation: ""}
+		maps.Copy(newAnnotations, network.GeneratePodAnnotations(vmi.Spec.Networks, vmi.Spec.Domain.Devices.Interfaces, pod.Annotations[networkv1.NetworkStatusAnnot], c.clusterConfig.GetNetworkBindings()))
+
 		if len(newAnnotations) != 0 {
```
Comment (Barakmor1) [inline pkg/virt-controller/watch/vmi.go:1278]: nit: i guess that this if is not required anymore

File: pkg/virt-controller/watch/migration.go [hunk 2]
```diff
@@ -515,6 +532,15 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
 	return nil
 }
 
+func (c *MigrationController) setMigrationFailed(migration *virtv1.VirtualMachineInstanceMigration) error {
+	migration.Status.Phase = virtv1.MigrationFailed
+	if _, exists := migration.GetAnnotations()[virtv1.EvacuationMigrationAnnotation]; !exists {
+		return nil
+	}
+
```
Comment (Barakmor1) [inline pkg/virt-controller/watch/migration.go:540]: Why do we need this if statement?

**Review by Barakmor1 — COMMENTED** [review]
Thanks @fossedihelm :)
left some comments but i still need to take a deeper look.

File: pkg/virt-controller/watch/vmi.go (hunk 1 above)
Comment (fossedihelm) [reply pkg/virt-controller/watch/vmi.go:1278]: Right! Thank you!

File: pkg/virt-controller/watch/migration.go (hunk 2 above)
Comment (fossedihelm) [reply pkg/virt-controller/watch/migration.go:540]: It is just because the `descheduler.alpha.kubernetes.io/eviction-in-progress` annotation is only added for evacuation migrations. There is no need to check the source virt-launcher pod annotations if the migration isn't an evacuation migration.
I am ready to drop this `if` you think that it does not worth it, but IMO this would save execution time and errors coming from reading the pod cache.

Comment (Barakmor1) [reply pkg/virt-controller/watch/migration.go:540]: I understand, but this code doesn't involve an API call, so the early exit won't notably improve performance. Therefore, I believe having less complexity is preferred in this case.

Comment (Barakmor1) [reply pkg/virt-controller/watch/migration.go:540]: I understand, but this code doesn't involve an API call, so the early exit won't notably improve performance. Therefore, I believe having less complexity is preferred in this case.

Comment (fossedihelm) [reply pkg/virt-controller/watch/migration.go]: Done

File: pkg/virt-controller/watch/vmi.go [hunk 3]
```diff
@@ -1271,14 +1272,14 @@ func (c *VMIController) sync(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod,
...
+		maps.Copy(newAnnotations, network.GeneratePodAnnotations(vmi.Spec.Networks, vmi.Spec.Domain.Devices.Interfaces, pod.Annotations[networkv1.NetworkStatusAnnot], c.clusterConfig.GetNetworkBindings()))
+
+		patchedPod, err := c.syncPodAnnotations(pod, newAnnotations)
+		if err != nil {
+			return &syncErrorImpl{err, FailedPodPatchReason}
 		}
+		*pod = *patchedPod
 
 		hotplugVolumes := getHotplugVolumes(vmi, pod)
 		hotplugAttachmentPods, err := controller.AttachmentPods(pod, c.podIndexer)
```
Comment (Barakmor1) [inline pkg/virt-controller/watch/vmi.go:1285]: @fossedihelm Shouldn't we handle the AttachmentPods as well?

also in migration: https://github.com/kubevirt/kubevirt/blob/841745da53a970eed302dccd12fdcce6f98a2edb/pkg/virt-controller/watch/migration.go#L412

File: staging/src/kubevirt.io/api/core/v1/types.go [hunk 4]
```diff
@@ -1055,6 +1055,13 @@ const (
 	// in which freePageReporting is always disabled.
 	FreePageReportingDisabledAnnotation string = "kubevirt.io/free-page-reporting-disabled"
 
+	// DeschedulerRequestEvictOnlyAnnotation indicates pods whose eviction is not expected to be completed right away.
+	// Instead, an eviction request is expected to be intercepted by an external component which will initiate the
+	// eviction process for the pod.
+	DeschedulerRequestEvictOnlyAnnotation string = "descheduler.alpha.kubernetes.io/request-evict-only"
+	// DeschedulerEvictionInProgressAnnotation indicates pods whose eviction was initiated by an external component.
+	DeschedulerEvictionInProgressAnnotation string = "descheduler.alpha.kubernetes.io/eviction-in-progress"
```
Comment (orelmisan) [inline staging/src/kubevirt.io/api/core/v1/types.go:1063]: This annotations affect the descheduler and not KubeVirt, so IMO they do not belong in KubeVirt's API.

**Review by orelmisan — COMMENTED** [review]
Thank you for the PR @fossedihelm.
I only skimmed through it and hope to have a deeper look later.

File: staging/src/kubevirt.io/api/core/v1/types.go (hunk 4 above)
Comment (fossedihelm) [reply staging/src/kubevirt.io/api/core/v1/types.go:1063]: That's right! Thanks

File: pkg/virt-controller/watch/migration.go [hunk 5]
```diff
@@ -1346,6 +1388,58 @@ func (c *MigrationController) sync(key string, migration *virtv1.VirtualMachineI
...
+func (c *MigrationController) removeDeschedulerAnnotationFromSourcePod(migration *virtv1.VirtualMachineInstanceMigration) error {
+	podKey := controller.NamespacedKey(migration.Namespace, migration.Status.MigrationState.SourcePod)
+	obj, exists, err := c.podIndexer.GetByKey(podKey)
+	if !exists {
+		log.Log.Reason(err).Errorf("source pod %s does not exist", migration.Status.MigrationState.SourcePod)
+		return nil
+	}
+	if err != nil {
+		log.Log.Reason(err).Errorf("Failed to fetch source pod %s for namespace from cache.", migration.Status.MigrationState.SourcePod)
+		return nil
```
Comment (Barakmor1) [inline pkg/virt-controller/watch/migration.go:1419]: shouldn't we return error here?

File: pkg/virt-controller/watch/migration.go [hunk 6]
```diff
@@ -1346,6 +1388,58 @@ func (c *MigrationController) sync(key string, migration *virtv1.VirtualMachineI
...
+	}
+
+	sourcePod := obj.(*k8sv1.Pod)
+	if _, exists := sourcePod.GetAnnotations()[virtv1.DeschedulerEvictionInProgressAnnotation]; exists {
+		patchSet := patch.New(
+			patch.WithTest(fmt.Sprintf("/metadata/annotations/%s", patch.EscapeJSONPointer(virtv1.DeschedulerEvictionInProgressAnnotation)), ""),
```
Comment (Barakmor1) [inline pkg/virt-controller/watch/migration.go:1425]: i'm not sure that the `WithTest` is needed here

File: pkg/virt-controller/watch/migration.go (hunk 5 above)
Comment (fossedihelm) [reply pkg/virt-controller/watch/migration.go:1419]: Right! Thank you!

File: pkg/virt-controller/watch/migration.go (hunk 6 above)
Comment (fossedihelm) [reply pkg/virt-controller/watch/migration.go:1425]: It is better to check that we are dropping something that exists.

File: pkg/virt-controller/watch/migration.go [hunk 7]
```diff
@@ -1346,6 +1390,62 @@ func (c *MigrationController) sync(key string, migration *virtv1.VirtualMachineI
...
+
+		_, err = c.clientset.CoreV1().Pods(sourcePod.Namespace).Patch(context.Background(), sourcePod.Name, types.JSONPatchType, patchBytes, v1.PatchOptions{})
+		if err != nil {
+			log.Log.Object(sourcePod).Errorf("failed to remove %s pod annotation : %v", deschedulerEvictionInProgressAnnotation, err)
+			return err
+		}
+	}
+
+	return nil
+}
```
Comment (jean-edouard) [inline pkg/virt-controller/watch/migration.go:1447]: Actually maybe the consts and these 2 functions could be moved to a new `descheduler.go`?
I find `migration.go` to be already big and complicated enough...

File: pkg/virt-controller/watch/migration_test.go [hunk 8]
```diff
@@ -1982,6 +2017,79 @@ var _ = Describe("Migration watcher", func() {
...
+				addPod(sourcePod)
+
+				controller.Execute()
+
+				updatedSourcePod, err := kubeClient.CoreV1().Pods(vmi.Namespace).Get(context.Background(), sourcePod.Name, metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				Expect(updatedSourcePod.Annotations).ToNot(HaveKeyWithValue(deschedulerEvictionInProgressAnnotation, ""))
+			})
+		})
+	})
```
Comment (jean-edouard) [inline pkg/virt-controller/watch/migration_test.go:2091]: Maybe also a `descheduler_test.go`?

File: pkg/virt-controller/watch/migration.go [hunk 9]
```diff
@@ -488,6 +514,9 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
 		migrationCopy.Status.Conditions = append(migrationCopy.Status.Conditions, condition)
 	} else if attachmentPodExists && podIsDown(attachmentPod) {
 		migrationCopy.Status.Phase = virtv1.MigrationFailed
+		if err := c.removeDeschedulerAnnotationFromSourcePod(migrationCopy); err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-controller/watch/migration.go:519]: This is soooo smoooth :man_facepalming: 
Thank you

File: pkg/virt-controller/watch/migration_test.go (hunk 8 above)
Comment (fossedihelm) [reply pkg/virt-controller/watch/migration_test.go:2090]: I would leave the test at the controller level, since the descheduler pkg works as an utility.
Also this would save us from a cycle dependency:
```
.-> //pkg/virt-controller/watch/descheduler:go_default_library 
`-- //pkg/virt-controller/watch:go_default_library 
```

**Comment by fossedihelm** [comment]
Thank you @jean-edouard for your review!
I addressed your comments, but one.
I also added some refactoring commits reducing the cycle dependency I found. Please let me know, thanks

File: pkg/virt-controller/watch/migration.go (hunk 6 above)
Comment (Barakmor1) [reply pkg/virt-controller/watch/migration.go]: You would use the `test` operation to make sure a condition is true before making changes. However, I’m not sure we need it here. Besides, we already checked that the annotation exists in the if condition. Could you please explain how the `test` operation improves this?

File: pkg/controller/controller.go [hunk 10]
```diff
@@ -46,6 +46,74 @@ const (
...
+	ImagePullBackOffReason = "ImagePullBackOff"
+	// NoSuitableNodesForHostModelMigration is set when a VMI with host-model CPU mode tries to migrate but no node
+	// is suitable for migration (since CPU model / required features are not supported)
+	NoSuitableNodesForHostModelMigration = "NoSuitableNodesForHostModelMigration"
+	// FailedPodPatchReason is set when a pod patch error occurs during sync
+	FailedPodPatchReason = "FailedPodPatch"
+	// MigrationBackoffReason is set when an error has occured while migrating
+	// and virt-controller is backing off before retrying.
+	MigrationBackoffReason = "MigrationBackoff"
+)
```
Comment (Barakmor1) [inline pkg/controller/controller.go:115]: IMO, it would be better to create a separate package just for functions and global variables. This way, you would reduce the risk of cyclic dependencies.

File: pkg/virt-controller/watch/migration.go (hunk 6 above)
Comment (fossedihelm) [reply pkg/virt-controller/watch/migration.go]: The test operation is commonly used to prevent an update when there's a concurrency conflict.
Even if we should be the only one removing this annotation, it can happen that the resource has been modified during the controller loop, hence that we are working with an older version. In this case the `test` operation allows us to be consistent, doing the check server side too.

File: pkg/controller/controller.go (hunk 10 above)
Comment (fossedihelm) [reply pkg/controller/controller.go:115]: In this case, I don't see it this way, since the `controller` pkg already exists and it fits perfectly to these constants (they are in the controller's scope). Creating a new pkg (I wouldn't even know what to call it honestly, since “controller” is already in use) gives no added value, as the controller package itself largely reduces the cyclic dependence, only importing external and KubeVirt API stuff.

File: pkg/virt-controller/watch/vmi.go (hunk 3 above)
Comment (fossedihelm) [reply pkg/virt-controller/watch/vmi.go:1149]: attachment pods need special handling that will be covered in a separate PR. Thanks for raising this! :+1:

Comment (Barakmor1) [reply pkg/virt-controller/watch/vmi.go:1149]: Great, can we open an issue regarding this to make sure it is not forgotten?

File: pkg/virt-controller/watch/migration.go (hunk 6 above)
Comment (Barakmor1) [reply pkg/virt-controller/watch/migration.go]: > Even if we should be the only one removing this annotation, it can happen that the resource has been modified during the controller loop, hence that we are working with an older version

why is that a problem?

This topic is important to me since the `WithTest` is used extensively in our project. I would really appreciate it if you could help me understand why it is necessary. 
Could you please explain why it would be problematic to call the patch API without the `WithTest` function even if the annotation is already deleted? Thank you!

Comment (iholder101) [reply pkg/virt-controller/watch/migration.go]: I definitely agree that `WithTest` is used more than it should.

I'm not sure about this case in particular, but I'd ask myself: "Should the call to `RemoveDeschedulerAnnotationFromSourcePod()` fail if the pod's descheduler annotation is already removed?"

Comment (fossedihelm) [reply pkg/virt-controller/watch/migration.go]: pod resource is not controlled only by us. From my point of view, we should remove the annotations that we've added. In this case in particular, the annotation is a descheduler one.
It should not happen, but if an external component edits the annotation we should not drop it, because we are going in concurrency with something else, and we should care about it.
Consider that the test checks not only the existence but the value too.

Using the test/remove vs remove can have different outputs:
- with test: we test that the annotation is present and has a specific value
- with plain remove: we drop it without checking the value.
So if the annotation is there but with a different value:
- with test: we fail
- with remove: we succeed
Basically, I am trying to cover the case where the annotation has been modified by something/someone else.

Maybe it is better to set a better value(in place of `""`)...
Having said that, I can remove the test, but I would ask first: why shouldn't we check that?
Also, from your point of view, in which case we should use test? Thank you

Comment (Barakmor1) [reply pkg/virt-controller/watch/migration.go]: Two points regarding this:

1. I don't think we should consider users who modify launcher annotations. If someone does so, it is their responsibility.

2. For the question about why we shouldn't check this: the test operation in patch should be used when we want to ensure we move from one phase to another. For instance, if the annotation looks like this:

```
phase: <phase_number>
```
and we want to move from phase 2 to phase 3, we need to make sure that the current phase is phase 2 and it hasn't changed to phase 4 or phase 1, for instance.
In this case, the test operation would make sense. This is a valid scenario where the test operation should be used, in my opinion.

Another use case would be to acquire a lock. For instance, let's say we have a CRD called "lock" with two fields: one for the lock owner and the second is a boolean that indicates if the lock is taken. In that case, when someone wants to acquire the lock, they should make sure it is not locked anymore before they modify the owner field.

Comment (fossedihelm) [reply pkg/virt-controller/watch/migration.go]: I cannot see too many differences actually. If the phase is checked programmatically it is the same case as this.
Anyway, since the majority are for dropping it I followed your comment.
Thanks for the discussion

File: pkg/virt-controller/watch/descheduler/descheduler.go [hunk 11]
```diff
@@ -19,7 +19,85 @@
...
+
 // EvictOnlyAnnotation indicates pods whose eviction is not expected to be completed right away.
 // Instead, an eviction request is expected to be intercepted by an external component which will initiate the
 // eviction process for the pod.
 const EvictOnlyAnnotation = "descheduler.alpha.kubernetes.io/request-evict-only"
+
+// EvictionInProgressAnnotation indicates pods whose eviction was initiated by an external component.
+const EvictionInProgressAnnotation = "descheduler.alpha.kubernetes.io/eviction-in-progress"
+
+func AddDeschedulerAnnotation(virtClient kubecli.KubevirtClient, sourcePod *k8sv1.Pod) error {
```
Comment (acardace) [inline pkg/virt-controller/watch/descheduler/descheduler.go:49]: nit: I'd call this `descheduler.MarkEvictionInProgress()', it comes from the descheduler package so I don't think you have to repeat that in the function name (at least in this case).

File: pkg/virt-controller/watch/descheduler/descheduler.go [hunk 12]
```diff
@@ -19,7 +19,85 @@
...
+	if _, exists := sourcePod.GetAnnotations()[EvictionInProgressAnnotation]; exists {
+		return nil
+	}
+
+	patchSet := patch.New(
+		patch.WithAdd(fmt.Sprintf("/metadata/annotations/%s", patch.EscapeJSONPointer(EvictionInProgressAnnotation)), ""),
```
Comment (acardace) [inline pkg/virt-controller/watch/descheduler/descheduler.go:55]: What happens if it exists already? Maybe a replace works better? The check you do before the patch doesn't guarantee 100% that the annotation won't exist as the Informer store might miss an update.

File: pkg/virt-controller/watch/descheduler/descheduler.go [hunk 13]
```diff
@@ -19,7 +19,85 @@
...
+		return nil
+	}
+	if err != nil {
+		log.Log.Reason(err).Errorf("Failed to fetch source pod %s for namespace from cache.", migration.Status.MigrationState.SourcePod)
+		return err
+	}
+
+	sourcePod := obj.(*k8sv1.Pod)
+	if _, exists := sourcePod.GetAnnotations()[EvictionInProgressAnnotation]; exists {
+		patchSet := patch.New(patch.WithRemove(fmt.Sprintf("/metadata/annotations/%s", patch.EscapeJSONPointer(EvictionInProgressAnnotation))))
```
Comment (acardace) [inline pkg/virt-controller/watch/descheduler/descheduler.go:89]: I'd do a test + remove for the same reason as above.

**Comment by jean-edouard** [comment]
/approve
My original comments were addressed and the refactoring looks great!
Just one thing I think could be improved is the release note. I needs to more user-focused, i.e. we don't break descheduler anymore.
Thank you!

File: pkg/virt-controller/watch/migration/migration.go [hunk 14]
```diff
@@ -1334,8 +1347,17 @@ func (c *MigrationController) sync(key string, migration *virtv1.VirtualMachineI
...
+			if err != nil {
+				return err
+			}
 		}
+
+		if migration.Status.Phase != virtv1.MigrationFailed {
+			return nil
+		}
+
+		return descheduler.RemoveDeschedulerAnnotationFromSourcePod(c.clientset, migration, c.podIndexer)
```
Comment (acardace) [inline pkg/virt-controller/watch/migration/migration.go:1360]: Shouldn't we remove the annotation only when the migration completes? Otherwise the descheduler might try to evict the source pod while is transferring the VM no?

**Comment by fossedihelm** [comment]
> /approve My original comments were addressed and the refactoring looks great! Just one thing I think could be improved is the release note. I needs to more user-focused, i.e. we don't break descheduler anymore. Thank you!

Thank you for the suggestion. Edited with: `Add descheduler compatibility`

File: pkg/virt-controller/services/template_test.go [hunk 15]
```diff
@@ -418,15 +418,16 @@ var _ = Describe("Template", func() {
...
+					v1.DomainAnnotation:                                  "testvmi",
+					"test":                                               "shouldBeInPod",
+					hooks.HookSidecarListAnnotationName:                  `[{"image": "some-image:v1", "imagePullPolicy": "IfNotPresent"}]`,
+					"pre.hook.backup.velero.io/container":                "compute",
+					"pre.hook.backup.velero.io/command":                  "[\"/usr/bin/virt-freezer\", \"--freeze\", \"--name\", \"testvmi\", \"--namespace\", \"testns\"]",
+					"post.hook.backup.velero.io/container":               "compute",
+					"post.hook.backup.velero.io/command":                 "[\"/usr/bin/virt-freezer\", \"--unfreeze\", \"--name\", \"testvmi\", \"--namespace\", \"testns\"]",
+					"kubevirt.io/migrationTransportUnix":                 "true",
+					"kubectl.kubernetes.io/default-container":            "compute",
+					"descheduler.alpha.kubernetes.io/request-evict-only": "",
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-controller/services/template_test.go:430]: https://github.com/kubevirt/kubevirt/pull/12025#discussion_r1631336380
:)

[the remark this answers is not in the document]
Comment (jean-edouard) [reply pkg/virt-controller/services/template_test.go:430]: That was actually a request from me: https://github.com/kubevirt/kubevirt/pull/12025#discussion_r1631336380

File: pkg/virt-controller/watch/descheduler/descheduler.go (hunk 11 above)
Comment (fossedihelm) [reply pkg/virt-controller/watch/descheduler/descheduler.go:49]: Nice suggestion, I moved it without renaming it. :+1:

File: pkg/virt-controller/watch/descheduler/descheduler.go (hunk 13 above)
Comment (fossedihelm) [reply pkg/virt-controller/watch/descheduler/descheduler.go:89]: And the discussion continues :) https://github.com/kubevirt/kubevirt/pull/12025#discussion_r1629204451

File: pkg/virt-controller/watch/descheduler/descheduler.go (hunk 12 above)
Comment (fossedihelm) [reply pkg/virt-controller/watch/descheduler/descheduler.go:55]: :+1:

**Comment by Barakmor1** [comment]
/lgtm
Thanks,  Great job!

**Comment by Barakmor1** [comment]
/hold 
Feel free to unhold once all comments are addressed

File: pkg/virt-controller/watch/descheduler/descheduler.go (hunk 12 above)
Comment (fossedihelm) [reply pkg/virt-controller/watch/descheduler/descheduler.go:55]: Actually, seems like `add` is better than `replace` https://jsonpatch.com/#replace.
But tested locally they work the same...

File: pkg/virt-controller/watch/migration/migration.go (hunk 14 above)
Comment (fossedihelm) [reply pkg/virt-controller/watch/migration/migration.go:1360]: If we reach this code it means that the migration fails, and `Failed` is a final status for migrations.
We add the annotation at the beginning and the source pod is annotated for the entire migration process.
In case the migration succeeds the source pod is completed and can remain annotated. In case 
the migration fails we drop the annotation.
IOW: this is under a switch case were the migration is `virtv1.MigrationPreparingTarget, virtv1.MigrationTargetReady, virtv1.MigrationFailed` and we early exit in case of:
```
if migration.Status.Phase != virtv1.MigrationFailed {
	return nil
}
```

**Comment by fossedihelm** [comment]
As requested:
1. Re-added test operation during annotation removal
1. Changed descheduler function names
1. Added `kubevirt` as value for the eviction in progress annotation

@acardace @Barakmor1 PTAL
Thank you!

File: pkg/virt-controller/watch/descheduler/descheduler.go (hunk 12 above)
Comment (acardace) [reply pkg/virt-controller/watch/descheduler/descheduler.go]: I just want to make sure it doesn't fail if the annotation is already there, does it?

File: pkg/virt-controller/watch/migration/migration.go (hunk 14 above)
Comment (acardace) [reply pkg/virt-controller/watch/migration/migration.go]: Sorry somehow the switch case tricked me, all good.

File: pkg/virt-controller/watch/descheduler/descheduler.go (hunk 12 above)
Comment (fossedihelm) [reply pkg/virt-controller/watch/descheduler/descheduler.go]: Nope, it does not. I tried with both kubectl and API request and everything is working.
i.e. This test:
```
			vmi := libvmifact.NewCirros()
			vmi = tests.RunVMIAndExpectLaunch(vmi, 60)
			patchBytes, err := patch.New(patch.WithAdd("/metadata/annotations/test", "false")).GeneratePayload()
			Expect(err).ToNot(HaveOccurred())

			vmi, err = virtClient.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, patchBytes, metav1.PatchOptions{})
			Expect(err).ToNot(HaveOccurred())

			patchBytes, err = patch.New(patch.WithAdd("/metadata/annotations/test", "last")).GeneratePayload()
			Expect(err).ToNot(HaveOccurred())

			vmi, err = virtClient.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, patchBytes, metav1.PatchOptions{})
			Expect(err).ToNot(HaveOccurred())
```
passes.

**Review by acardace — APPROVED** [review]

**Comment by brianmcarey** [comment]
@fossedihelm @acardace I'm not sure this will merge with the coveralls check failing.

**Comment by fossedihelm** [comment]
> @fossedihelm @acardace I'm not sure this will merge with the coveralls check failing.

Actually I don't know why this causes this coveralls coverage decrease...

**Comment by fossedihelm** [comment]
Added controller unit test, trying to help coverall

**Comment by acardace** [comment]
The huge decrease in coverage is due to constants and utility functions being moved out of vmi.go, there is no actual change in code coverage.

I don't think this failure should block this PR, it really has no value.

@xpivarc, @brianmcarey if you agree I'd override the the coveralls lane for this one time.

**Comment by EdDev** [comment]
> The huge decrease in coverage is due to constants and utility functions being moved out of vmi.go, there is no actual change in code coverage.
> 
> I don't think this failure should block this PR, it really has no value.
> 
> @xpivarc, @brianmcarey if you agree I'd override the the coveralls lane for this one time.

Reduction of coverage hints about the potential degradation of code-quality. 1.3% is a lot.
If code moved out and lost coverage, it is a real problem. This usually happens when coverage is implicit and not explicit. But implicit is better than none.

Perhaps you should reconsider the refactoring part and split it to a follow up PR.

**Comment by fossedihelm** [comment]
Looks like https://github.com/kubevirt/kubevirt/pull/12100 coverage is not complaining, actually is increasing...
Basically it drops the https://github.com/kubevirt/kubevirt/pull/12025/commits/a7556cdc7cf1676af8ce2d834e8ff504e1a79480 commit

**Comment by acardace** [comment]
Go figure, looks like the tool just got in the way this time around.
